### PR TITLE
[SPARK-5088] Use spark-class for running executors directly

### PR DESCRIPTION
**Pull request by jongyoul**

### CODE REVIEW

**Comment by jongyoul** [comment]
@tnachen On the contrary, why mesos has two different launcher? I couldn't find any idea in a history of that codes, and spark-executor calls spark-class which has a case statement internally. I think that merging two similar codes is a good idea in maintaining codes.

**Comment by tnachen** [comment]
I'm not part of that history either, I just want to know the motivation since the PR doesn't say anything.
For code cleanup it does make sense. 
Your patch looks good to me, unfortunately we need to find some committer to look at this now....

**Comment by jongyoul** [comment]
@tnachen I'm sorry for not having any comment. I think that this PR is very clear to guess my opinion.

@pwendell Please review this PR which makes some mesos codes clean. Because this PR is about maintaining code, I ask you to review this PR.

**Comment by jongyoul** [comment]
@JoshRosen @tgravescs @andrewor14 Could anyone review this PR? That makes mesos codes clean

**Comment by jongyoul** [comment]
Please review this PR

**Comment by mateiz** [comment]
This looks okay, but is there any motivation for it other than getting rid of that source file?

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala [hunk 1]
```diff
@@ -118,14 +119,16 @@ private[spark] class MesosSchedulerBackend(
     val command = CommandInfo.newBuilder()
       .setEnvironment(environment)
     val uri = sc.conf.get("spark.executor.uri", null)
+    val executorBackendName = classOf[MesosExecutorBackend].getName
     if (uri == null) {
-      val executorPath = new File(executorSparkHome, "/sbin/spark-executor").getCanonicalPath
+      val executorPath = new File(executorSparkHome, s"/bin/spark-class $executorBackendName")
+        .getCanonicalPath
```
Comment (mateiz) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: This is actually weird, why are we including `$executorBackendName` in the filename? We should just look for spark-class. I'm not even sure getCanonicalPath will work on a file that doesn't exist

**Comment by jongyoul** [comment]
@mateiz That's all, but I think it is meaningful to remove some redundant codes and helps understand codes easy. And I wonder why that code exists. I think 'spark-executor`is just a wrapper for launching `spark-class`. And we need to try to merge coarse and fine grained mode. That's the first and easiest effort for it

File: core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala (hunk 1 above)
Comment (jongyoul) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerBackend.scala]: That's my mistake. I'll fix it right now

**Comment by mateiz** [comment]
Coarse and fine-grained mode are different on purpose, I don't think it makes sense to merge those. BTW have you tested both of them on this?

**Comment by jongyoul** [comment]
@mateiz I've test it both mode.

**Comment by jongyoul** [comment]
@mateiz I've fixed weird codes by your comments and added test case whether `spark.executor.uri` exists or not.

**Comment by jongyoul** [comment]
@mateiz Could you please explain the purposes of coarse and fine-grained mode shortly? It looks similar. The only difference between two modes is the way to offer resources. I have no idea of history of existing two modes.

**Comment by jongyoul** [comment]
retest this please.I don't know why that simple test fails.

**Comment by jongyoul** [comment]
@mateiz Passed tests. Please review it again.

**Comment by jongyoul** [comment]
@mateiz Check this PR finally. I think It would be good by your review.

**Comment by mateiz** [comment]
@jongyoul sorry for the delayed reply, but this looks good. Mind rebasing it on master though? It's now fallen behind.

**Comment by jongyoul** [comment]
@mateiz I'll rebase this PR from master.

**Comment by jongyoul** [comment]
Rebase is not finished.

**Comment by jongyoul** [comment]
retest this please.

**Comment by jongyoul** [comment]
@mateiz I've rebased this PR and finished tests successfully. Merge this, please.

**Comment by pwendell** [comment]
LGTM I will pull it in.

**Comment by mateiz** [comment]
Thanks, looks good!

**Pull request closed without merging**